File: common/src/main/java/com/ultreon/mods/advanceddebug/client/menu/DebugRenderContext.java
package com.ultreon.mods.advanceddebug.client.menu;

import com.ultreon.mods.advanceddebug.api.client.menu.IDebugRenderContext;
import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public abstract class DebugRenderContext implements IDebugRenderContext {
    private int left;
    private int top;
    private int right;

    private final GuiGraphics guiGraphics;
    private final Minecraft mc = Minecraft.getInstance();
    private final int width;
    private final int height;
    private static final int HORIZONTAL_OFFSET = 6;
    private static final int VERTICAL_OFFSET = 6;

    public DebugRenderContext(@NotNull GuiGraphics gfx, int width, int height) {
        this.guiGraphics = gfx;
        this.width = width;
        this.height = height;
    }

    public GuiGraphics getGuiGraphics() {
        return this.guiGraphics;
    }

    public int getWidth() {
        return this.width;
    }

    public int getHeight() {
        return this.height;
    }

    @Override
    public void left(Component text, Object object, Object... objects) {
        this.drawLeft(this.guiGraphics, text.getString(), this.left++, object, objects);
    }

    @Override
    public void left(String text, Object object, Object... objects) {
        this.drawLeft(this.guiGraphics, text, this.left++, object, objects);
    }

    @Override
    public void left(Component text) {
        this.drawLeft(this.guiGraphics, text.getString(), this.left++);
    }

    @Override
    public void left(String text) {
        this.drawLeft(this.guiGraphics, ChatFormatting.GRAY + "-== " + text + ChatFormatting.GRAY + " ==-", this.left++);
    }

    @Override
    public void left() {
        this.left++;
    }

    @Override
    public void right(Component text, Object object, Object... objects) {
        this.drawRight(this.guiGraphics, text.getString(), this.right++, object, objects);
    }

    @Override
    public void right(String text, Object object, Object... objects) {
        this.drawRight(this.guiGraphics, text, this.right++, object, objects);
    }

    @Override
    public void right(Component text) {
        this.drawRight(this.guiGraphics, text.getString(), this.right++);
    }

    @Override
    public void right(String text) {
        this.drawRight(this.guiGraphics, ChatFormatting.GRAY + "-== " + text + ChatFormatting.GRAY + " ==-", this.right++);
    }

    @Override
    public void right() {
        this.right++;
    }

    @Override
    public void top(Component text) {
        this.drawTop(this.guiGraphics, text.getString(), this.top++);
    }

    @Override
    public void top(String text) {
        this.drawTop(this.guiGraphics, text, this.top++);
    }

    @Override
    public void top() {
        this.top++;
    }

    private void drawTop(@NotNull GuiGraphics gfx, String text, int line) {
        this.drawLine(gfx, Component.literal(text), (int) (this.width / 2f - this.mc.font.width(text) / 2f), VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawTop(@NotNull GuiGraphics gfx, Component text, int line) {
        this.drawLine(gfx, text, (int) (this.width / 2f - this.mc.font.width(text) / 2f), VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawLeft(@NotNull GuiGraphics gfx, String text, int line, Object obj, Object... objects) {
        this.drawLine(gfx, DebugGui.get().format(text, obj, objects), HORIZONTAL_OFFSET, VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawLeft(@NotNull GuiGraphics gfx, String text, int line) {
        this.drawLine(gfx, Component.literal(text), HORIZONTAL_OFFSET, VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawLeft(@NotNull GuiGraphics gfx, Component text, int line) {
        this.drawLine(gfx, text, HORIZONTAL_OFFSET, VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawRight(@NotNull GuiGraphics gfx, String text, int line, Object obj, Object... objects) {
        Component format = DebugGui.get().format(text, obj, objects);
        this.drawLine(gfx, format, this.width - HORIZONTAL_OFFSET - this.mc.font.width(format), VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawRight(@NotNull GuiGraphics gfx, String text, int line) {
        this.drawLine(gfx, Component.literal(text), this.width - HORIZONTAL_OFFSET - this.mc.font.width(text), VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    private void drawRight(@NotNull GuiGraphics gfx, Component text, int line) {
        this.drawLine(gfx, text, this.width - HORIZONTAL_OFFSET - this.mc.font.width(text), VERTICAL_OFFSET + line * (this.mc.font.lineHeight + 2));
    }

    protected abstract void drawLine(@NotNull GuiGraphics gfx, Component text, int x, int y);

    public void left(GuiGraphics gfx, MutableComponent component) {
        this.drawLeft(gfx, component, this.left++);
    }

    public void right(GuiGraphics gfx, MutableComponent component) {
        this.drawRight(gfx, component, this.right++);
    }

    public void top(GuiGraphics gfx, MutableComponent component) {
        this.drawRight(gfx, component, this.top++);
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/client/menu/DebugRenderContext.kt
package com.ultreon.mods.advanceddebug.client.menu

import com.mojang.blaze3d.vertex.PoseStack
import com.ultreon.mods.advanceddebug.api.client.menu.IDebugRenderContext
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component

abstract class DebugRenderContext(private val pose: PoseStack, val width: Int, val height: Int) : IDebugRenderContext {
    private var left = 0
    private var top = 0
    private var right = 0
    private val mc = Minecraft.getInstance()

    override fun left(text: Component, `object`: Any?, vararg objects: Any?) {
        drawLeft(pose, text.string, left++, `object`, *objects)
    }

    override fun left(text: String, `object`: Any?, vararg objects: Any?) {
        drawLeft(pose, text, left++, `object`, *objects)
    }

    override fun left(text: Component) {
        drawLeft(pose, text.string, left++)
    }

    override fun left(text: String) {
        drawLeft(pose, ChatFormatting.GRAY.toString() + "-== " + text + ChatFormatting.GRAY + " ==-", left++)
    }

    override fun left() {
        left++
    }

    override fun right(text: Component, `object`: Any?, vararg objects: Any?) {
        drawRight(pose, text.string, right++, `object`, *objects)
    }

    override fun right(text: String, `object`: Any?, vararg objects: Any?) {
        drawRight(pose, text, right++, `object`, *objects)
    }

    override fun right(text: Component) {
        drawRight(pose, text.string, right++)
    }

    override fun right(text: String) {
        drawRight(pose, ChatFormatting.GRAY.toString() + "-== " + text + ChatFormatting.GRAY + " ==-", right++)
    }

    override fun right() {
        right++
    }

    override fun top(text: Component) {
        drawTop(pose, text.string, top++)
    }

    override fun top(text: String) {
        drawTop(pose, text, top++)
    }

    override fun top() {
        top++
    }

    private fun drawTop(pose: PoseStack, text: String, line: Int) {
        drawLine(
            pose,
            Component.literal(text),
            (width / 2f - mc.font.width(text) / 2f).toInt(),
            VERTICAL_OFFSET + line * (mc.font.lineHeight + 2)
        )
    }

    private fun drawLeft(pose: PoseStack, text: String, line: Int, obj: Any?, vararg objects: Any?) {
        drawLine(
            pose,
            DebugGui.get().format(text, obj, *objects),
            HORIZONTAL_OFFSET,
            VERTICAL_OFFSET + line * (mc.font.lineHeight + 2)
        )
    }

    private fun drawLeft(pose: PoseStack, text: String, line: Int) {
        drawLine(pose, Component.literal(text), HORIZONTAL_OFFSET, VERTICAL_OFFSET + line * (mc.font.lineHeight + 2))
    }

    private fun drawRight(pose: PoseStack, text: String, line: Int, obj: Any?, vararg objects: Any?) {
        val format: Component = DebugGui.get().format(text, obj, *objects)
        drawLine(
            pose,
            format,
            width - HORIZONTAL_OFFSET - mc.font.width(format),
            VERTICAL_OFFSET + line * (mc.font.lineHeight + 2)
        )
    }

    private fun drawRight(pose: PoseStack, text: String, line: Int) {
        drawLine(
            pose,
            Component.literal(text),
            width - HORIZONTAL_OFFSET - mc.font.width(text),
            VERTICAL_OFFSET + line * (mc.font.lineHeight + 2)
        )
    }

    protected abstract fun drawLine(pose: PoseStack, text: Component, x: Int, y: Int)

    companion object {
        private const val HORIZONTAL_OFFSET = 6
        private const val VERTICAL_OFFSET = 6
    }
}